subsys/power/device_pm.c

```diff
@@ -122,25 +126,19 @@ static int device_pm_request(const struct device *dev,
 		return 0;
 	}
 
-	k_work_submit(&dev->pm->work);
+	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
 
 	/* Return in case of Async request */
 	if (pm_flags & DEVICE_PM_ASYNC) {
 		return 0;
 	}
 
-	/* Incase of Sync request wait for completion event */
-	do {
-		(void)k_poll(&dev->pm->event, 1, K_FOREVER);
-		k_poll_signal_check(&dev->pm->signal,
-						&signaled, &result);
-	} while (!signaled);
-
-	dev->pm->event.state = K_POLL_STATE_NOT_READY;
-	k_poll_signal_reset(&dev->pm->signal);
-
+	k_mutex_init(&request_mutex);
+	k_mutex_lock(&request_mutex, K_FOREVER);
+	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
+	k_mutex_unlock(&request_mutex);
 
-	return result == target_state ? 0 : -EIO;
+	return target_state == atomic_get(&dev->pm->fsm_state) ? 0 : -EIO;
```

Comment:
This code is a bit hard to understand. How does it actually end up being in the target state? How does it know what to do, since target_state is not stored anywhere.

Reply:
you request to device to change to `target_state`, as this operation may be asynchronous the device is responsible to call the given callback with its state after this operation. The state given by the device is then stored in `dev->pm->fsm_state` we then compare here to check if the state was properly set.